proxy/coco/tests/replication.rs

```diff
@@ -215,3 +247,75 @@ async fn can_fetch_project_changes() -> Result<(), Box<dyn std::error::Error>> {
 
     Ok(())
 }
+
+#[tokio::test]
+async fn can_sync_on_startup() -> Result<(), Box<dyn std::error::Error>> {
+    init_logging();
+
+    let alice_tmp_dir = tempfile::tempdir()?;
+    let alice_repo_path = alice_tmp_dir.path().join("radicle");
+    let alice_store = kv::Store::new(kv::Config::new(alice_tmp_dir.path().join("store")))?;
+    let (alice_peer, alice_state, alice_signer) = build_peer(&alice_tmp_dir).await?;
+    let alice_addr = alice_state.lock().await.listen_addr();
+    let alice_peer_id = alice_state.lock().await.peer_id();
+    let alice_events = alice_peer.subscribe();
+
+    let bob_tmp_dir = tempfile::tempdir()?;
+    let bob_store = kv::Store::new(kv::Config::new(bob_tmp_dir.path().join("store")))?;
+    let (bob_peer, bob_state, bob_signer) = build_peer_with_seeds(
+        &bob_tmp_dir,
+        vec![Seed {
+            addr: alice_addr,
+            peer_id: alice_peer_id.clone(),
+        }],
+    )
+    .await?;
+    let bob_peer_id = bob_state.lock().await.peer_id();
+
+    let alice = {
+        let alice_signer = alice_signer.clone();
+        let alice_state = alice_state.clone();
+        let ally = alice_state.lock_owned().await;
+        tokio::task::spawn_blocking(move || ally.init_owner(&alice_signer.clone(), "alice"))
+            .await??
+    };
+    let _bob = {
+        let bob_state = bob_state.clone();
+        let bobby = bob_state.lock_owned().await;
+        tokio::task::spawn_blocking(move || bobby.init_owner(&bob_signer, "bob")).await??
+    };
+    {
+        let alice = alice.clone();
+        let alice_repo_path = alice_repo_path.clone();
+        let alice_state = alice_state.clone();
+        let ally = alice_state.lock_owned().await;
+        tokio::task::spawn_blocking(move || {
+            ally.init_project(
+                &alice_signer,
+                &alice,
+                &shia_le_pathbuf(alice_repo_path.clone()),
+            )
+        })
+        .await??;
+    };
+
+    {
+        let bob_events = bob_peer.subscribe();
+        tokio::task::spawn(alice_peer.run(
+            alice_state.clone(),
+            alice_store,
+            RunConfig {
+                sync_on_startup: true,
+            },
+        ));
+        tokio::task::spawn(bob_peer.run(bob_state.clone(), bob_store, RunConfig::default()));
+        connected(bob_events, &alice_peer_id).await?;
+    };
+
+    assert_event!(
```

Comment:
I also expected an assertion that bob has the updates in his `Storage` after syncing up with alice

Reply:
I didn't go through the length of setting up a project both peers track and that has new changes since last being online. That's mostly because I dunno how to disconnect and reconnect a peer.

Reply:
Hmmm, I wonder if we create and drop the peer in a block, then create it again outside of that block would it be equivalent?

Reply:
For that we have to have a way to hold onto it's monorepo/start and assoicated key/signer.

Reply:
We could create the key outside of the block and initialise the storage inside and outside the block. So in pseudo-code:
```rust
let key = SecretKey::new();

{
  let bob = init_peer(key.clone());
  let repo = bob.clone_repo(urn);
}

let commit = alice.commit();
let bob = init_peer(key.clone());
wait_for_sync(bob);
assert!(bob.has_commit(comit))
```

Reply:
fyi: I'd really like to see this test. Are we going to cover it in another PR or are you going to cover it here? It's what I'm holding out on before I give a :+1: :)

Reply:
It's under way, but quite involved. Our test suite is not at a point where these kind of scenarios can be expressed concisely, as a result the test code feels more like a liability than a confidence improvement.

Reply:
Ok, maybe we should tackle it as a next step then?

Reply:
https://github.com/radicle-dev/radicle-upstream/issues/908#issuecomment-696020816